oauth2_provider/contrib/rest_framework/authentication.py

```diff
@@ -23,10 +24,18 @@ def authenticate(self, request):
         Returns two-tuple of (user, token) if authentication succeeds,
         or None otherwise.
         """
+        if request is None:
+            return None
         oauthlib_core = get_oauthlib_core()
-        valid, r = oauthlib_core.verify_request(request, scopes=[])
-        if valid:
-            return r.user, r.access_token
+        try:
+            valid, r = oauthlib_core.verify_request(request, scopes=[])
+        except ValueError as error:
+            if str(error) == "Invalid hex encoding in query string.":
```

Comment:
Is there a way to detect this case other than a string compare of the error message text? Not sure if oauthlib supports international languages but I could see this possibly being an issue. This feels a bit fragile.

Reply:
FWIW, the string is hard-coded as the error message in oauthlib: https://github.com/oauthlib/oauthlib/blob/f9a07c6c07d0ddac255dd322ef5fc54a7a46366d/oauthlib/common.py#L122

There are also other ValueErrors it might throw, I'm not sure if they merit similar treatment: https://github.com/oauthlib/oauthlib/blob/f9a07c6c07d0ddac255dd322ef5fc54a7a46366d/oauthlib/common.py#L115

Reply:
@sarayourfriend Would throwing any ValueError be sufficient to raise SuspiciousOperation then?

Reply:
Additionally, this is the exactly logic that's already present in `backends.py`, just applied to the rest framework extension!

https://github.com/jazzband/django-oauth-toolkit/blob/d35f030960617cb4d0dbe9a3e89b797df2e7cf0c/oauth2_provider/backends.py#L18-L24

If you'd prefer we raise a `SuspiciousOperation` on all `ValueError`s, then I can also apply that change there 🙂

Reply:
@n2ygk I can't say either way, I'm not familiar with the intention behind the SuspiciousOperation exception, nor with all the code paths in `verify_request` or how those might interact with the points of extensibility that exist in oauthlib. What's the intention behind SuspiciousOperation?

Whilst matching the string in the exception doesn't feel good, I agree with @AetherUnbound that it's the already existing method for disambiguating exceptions, and your original concern about i18n support in the library doesn't appear to apply. I'd vote to keep the targeted, string comparison approach, and perhaps open an issue upstream if this library thinks a more stable method of disambiguation would be better (like subclasses of `ValueError` with some enumerated attribute to discriminate the cause).